src/PeNet/Header/Resource/Icon.cs

```diff
@@ -40,132 +41,69 @@ public Icon(IRawFile peFile, long offset, uint size, uint id, Resources parent)
         /// <summary>
         ///     Byte span of the icon image.
         /// </summary>
-        public Span<byte> AsSpan()
+        public Span<byte> AsRawSpan()
         {
             return PeFile.AsSpan(Offset, Size);
         }
 
+        private bool IsPng => AsRawSpan().Slice(0, 8).SequenceEqual(PNGHeader);
+        private bool IsIco => !IsPng && AssociatedGroupIconDirectoryEntry is not null && !AsRawSpan().IsEmpty;
+
         /// <summary>
         ///     Adding .ICO-Header to the bytes of the icon image.
         /// </summary>
         /// <returns>Bytes of the icon image as .ICO file.</returns>
-        public byte[]? AsICO()
+        public byte[]? AsIco()
         {
-            if (AssociatedGroupIconDirectoryEntry == null) return null;
-            var iconBytes = AsSpan().ToArray();
-            if (iconBytes == null) return null;
+            var raw = AsRawSpan();
+            if (IsPng) return raw.ToArray(); // TODO: CFF does not use an additional header for .PNG. But this would not break anything.
+            if (!IsIco) return null;
 
-            if (iconBytes.Take(8).SequenceEqual(PNGHeader)) return iconBytes; // TODO: CFF does not use an additional header for .PNG. But this would not break anything.
+            var icoBytes = new byte[IcoHeaderSize + raw.Length];
+            var icoHeader = IcoHeader();
 
-            var iconICOBytes = new byte[ICOHeaderSize + iconBytes.Length];
-            SetICOHeader(iconICOBytes);
-            
-            iconBytes.CopyTo(iconICOBytes, ICOHeaderSize);
-            return iconICOBytes;
+            icoHeader.AsSpan().CopyTo(new Span<byte>(icoBytes, 0, (int) IcoHeaderSize));
+            raw.CopyTo(new Span<byte>(icoBytes, (int) IcoHeaderSize, raw.Length));
+            return icoBytes;
         }
 
-        /// <summary>
-        ///     Setting .ICO-Header with data from the associated GroupIconDirectoryEntry.
-        ///     According to the structure described in :https://en.wikipedia.org/wiki/ICO_(file_format)
-        /// </summary>
-        /// <param name="iconICOBytes">Byte array to process.</param>
-        private void SetICOHeader(IList<byte> iconICOBytes)
+        private byte[] IcoHeader()
         {
-            SetIconDirectoryStructure(iconICOBytes);
-            SetIconDirectoryEntryStructure(iconICOBytes);
+            var stream = new MemoryStream((int) IcoHeaderSize);
+            WriteIconDirectoryStructure(stream);
+            WriteIconDirectoryEntryStructure(stream);
+            return stream.ToArray();
         }
 
-        /// <summary>
-        ///     Setting bytes in IconDirectoryStructure.
-        /// </summary>
-        /// <param name="iconICOBytes">Byte array to process.</param>
-        private static void SetIconDirectoryStructure(IList<byte> iconICOBytes)
+        private static void WriteIconDirectoryStructure(Stream stream)
```

Comment:
This does not seem more readable to me.
I really miss the matchability with the header standard regarding positions.

Reply:
What's the issue? I think it translates pretty well: https://docs.fileformat.com/image/ico/#header
Especially since the file format explicitly states that [little endian is used](https://docs.fileformat.com/image/ico/#ico-file-format)